crates/proto/src/rr/rdata/cert.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2024 Brian Taber <btaber@zsd.systems>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// https://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// https://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! CERT record type for storing certificates in DNS
+use std::fmt;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::*,
+    rr::{RData, RecordData, RecordDataDecodable, RecordType},
+    serialize::binary::*,
+};
+
+/// [RFC 4398, Storing Certificates in DNS, November 1987][rfc4398]
+/// https://tools.ietf.org/html/rfc4398
+///
+/// ```text
+///
+/// The CERT resource record (RR) has the structure given below.  Its RR
+/// type code is 37.
+///
+///    1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+/// 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |             type              |             key tag           |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |   algorithm   |                                               /
+/// +---------------+            certificate or CRL                 /
+/// /                                                               /
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-|
+/// ```
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[derive(Default, Debug, PartialEq, Eq, Hash, Clone)]
+pub struct CERT {
+    cert_type: u32,
+    key_tag: u32,
+    algorithm: u32,
+    cert_data: Vec<u8>,
+}
+
+impl CERT {
+    /// Construct a new CERT RData
+    pub const fn new() -> Self {
+        Self {
+            cert_type: 0,
+            key_tag: 0,
+            algorithm: 0,
+            cert_data: vec![],
+        }
+    }
+
+    /// Constructs a new CERT RData with the associated data
+    pub fn with(cert_data: Vec<u8>) -> Self {
+        let cert_type: u32 = ((cert_data[0] as u32) << 8) | (cert_data[1] as u32);
+        let key_tag: u32 = ((cert_data[2] as u32) << 8) | (cert_data[3] as u32);
+        let algorithm: u32 = cert_data[4].into();
+        let cert_data: Vec<u8> = cert_data[5..].to_vec();
```

Comment:
These array indexes could panic if this function is passed with a cert_data smaller than 6 elements.

Reply:
added tests/check to return error